compiler/src/iree/compiler/Codegen/Common/LinkTuningSpecsPass.cpp

```diff
@@ -61,6 +62,210 @@ static bool consumesInputOp(NamedSequenceOp op) {
+    if (auto namedSequenceOp = dyn_cast<NamedSequenceOp>(parentOp)) {
+      return namedSequenceOp.getArgAttr(blockArg.getArgNumber(),
+                                        kArgConsumedAttrName) != nullptr;
+    }
+  }
+  return false;
+}
+
+struct TuningSpecsToMerge {
+  // - `namedSequenceOpsToMove`: Contains `NamedSequenceOp`s that either:
+  //     - Are not named `__kernel_config`.
+  //     - Do not have the `iree_codegen.tuning_spec_entrypoint` attribute.
+  SmallVector<NamedSequenceOp> namedSequenceOpsToMove;
+  // Maps a `NamedSequenceOp` to its user `ForeachMatchOp`, in which the
+  // `NamedSequenceOp` is used as a matcher or action inside a
+  // `ForeachMatchOp`.
+  llvm::DenseMap<NamedSequenceOp, ForeachMatchOp> namedSequenceToUser;
+};
+
+// Populates the mapping of `NamedSequenceOp` to `ForeachMatchOp`
+// by processing matchers and actions inside the given `ForeachMatchOp`.
+//
+// This function iterates over all matcher and action references in
+// `foreachMatch`. If a reference is a valid `SymbolRefAttr` that resolves to a
+// `NamedSequenceOp` within the `module`, the function records the association
+// by mapping the `NamedSequenceOp` to `foreachMatch` in `namedSequenceToUser`.
+static void updateForeachMatchMappings(
+    ForeachMatchOp foreachMatch, ModuleOp module,
+    llvm::DenseMap<NamedSequenceOp, ForeachMatchOp> &namedSequenceToUser) {
+  for (auto matcher : foreachMatch.getMatchers()) {
+    if (auto matcherSymRef = mlir::dyn_cast<mlir::SymbolRefAttr>(matcher)) {
+      if (auto matcherOp =
+              SymbolTable::lookupNearestSymbolFrom<NamedSequenceOp>(
+                  module, matcherSymRef)) {
+        namedSequenceToUser[matcherOp] = foreachMatch;
+      }
+    }
+  }
+
+  for (auto action : foreachMatch.getActions()) {
+    if (auto actionSymRef = mlir::dyn_cast<mlir::SymbolRefAttr>(action)) {
+      if (auto actionOp = SymbolTable::lookupNearestSymbolFrom<NamedSequenceOp>(
+              module, actionSymRef)) {
+        namedSequenceToUser[actionOp] = foreachMatch;
+      }
+    }
+  }
+}
+
+// Collects tuning specs (`NamedSequenceOp`s) from all inner modules within the
+// given input model, and organizes them into a `TuningSpecsToMerge` struct.
+static TuningSpecsToMerge collectTuningSpecsToMerge(ModuleOp module) {
+  TuningSpecsToMerge tuningSpecs;
+  for (auto innerModule : module.getBody()->getOps<ModuleOp>()) {
+    for (auto namedSequenceOp :
+         innerModule.getBody()->getOps<NamedSequenceOp>()) {
+      if (namedSequenceOp.getSymName() != kKernelConfigSpecName ||
+          !namedSequenceOp->hasAttr(kTuningSpecEntrypointAttrName)) {
+        tuningSpecs.namedSequenceOpsToMove.push_back(namedSequenceOp);
+        continue;
+      }
+
+      auto foreachMatchOps = namedSequenceOp.getOps<ForeachMatchOp>();
+      ForeachMatchOp foreachMatch = getSingleElement(foreachMatchOps);
+      assert(foreachMatch &&
+             "ForeachMatch should exist in the `__kernel_config`.");
```

Comment:
I don't think this assertion will ever trigger -- if there are no foreach match ops, the list will be empty, not filled with nullptr values